### PR TITLE
Support spark dataframe containing array or object

### PR DESCRIPTION
**Pull request by harupy**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Depends on #9943?

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/spark_udf/complex.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import random
```
Comment (dbczumar) [inline examples/spark_udf/complex.py:1]: @harupy Nit: Can we adjust the name of the example? Not sure if we want to tell users that our example is "complex"

Comment (harupy) [reply examples/spark_udf/complex.py]: Renamed!

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -906,11 +906,12 @@ def _enforce_datatype(data: Any, dtype: DataType):
 def _enforce_array(data: Any, arr: Array, required=True):
     if not required and data is None:
         return None
-    if not isinstance(data, list):
+    if not isinstance(data, (list, np.ndarray)):
         if np.isscalar(data):
             data = [data]
         else:
             raise MlflowException(f"Expected data to be list, got {type(data).__name__}")
+        return data
```
Comment (harupy) [inline mlflow/models/utils.py:914]: @serena-ruan Is `return` missing here?

Comment (harupy) [reply mlflow/models/utils.py:914]: Never mind. We have multiple if statements.

File: mlflow/models/utils.py [hunk 3]
```diff
@@ -947,11 +947,12 @@ def _enforce_object(data: Dict[str, Any], obj: Object, required=True):
         raise MlflowException(
             "Invalid properties not defined in the schema found: " f"{invalid_props}"
         )
+    # raise Exception(data, properties)
```
Comment (harupy) [inline mlflow/models/utils.py:950]:
```suggestion
```

File: mlflow/models/utils.py [hunk 4]
```diff
@@ -947,11 +947,12 @@ def _enforce_object(data: Dict[str, Any], obj: Object, required=True):
...
     for k, v in data.items():
         try:
             data[k] = _enforce_property(v, properties[k])
         except MlflowException as e:
-            raise MlflowException(f"Failed to enforce schema for key `{k}`") from e
+            raise MlflowException(f"Failed to enforce schema for key `{k}`: {e}") from e
```
Comment (harupy) [inline mlflow/models/utils.py:955]:
```suggestion
            raise MlflowException(f"Failed to enforce schema for key `{k}`") from e
```

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -1377,7 +1369,15 @@ def _predict_row_batch(predict_fn, args):
                         " (Since the columns were passed unnamed they are expected to be in"
                         " the order specified by the schema).".format(len(names), names, len(args))
                     )
-            pdf = pandas.DataFrame(data={names[i]: x for i, x in enumerate(args)}, columns=names)
+            pdf = pandas.DataFrame(
+                data={
+                    names[i]: x
+                    if isinstance(x, pandas.Series)
+                    else x.apply(lambda row: row.to_dict(), axis=1)
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:1376]: Can we add a comment here to explain this is for coerce a pandas DataFrame into a pandas Series for StructType support?

File: tests/pyfunc/test_spark.py [hunk 6]
```diff
@@ -1298,3 +1298,59 @@ def predict(self, context, model_input, params=None):
...
+    )
+    mlflow.pyfunc.save_model(
+        path=tmp_path,
+        python_model=MyModel(),
+        signature=mlflow.models.infer_signature(df),
+    )
+
+    udf = mlflow.pyfunc.spark_udf(spark=spark, model_uri=tmp_path, result_type="string")
+    pdf = df.withColumn("output", udf("str", "arr", "obj", "obj_arr")).toPandas()
+    assert pdf["output"][0] == "a | [0] | {'bool': True} | [{'double': 0.1}]"
```
Comment (serena-ruan) [inline tests/pyfunc/test_spark.py:1356]: This looks good! Can we add more tests for more complex structures and more rows?

File: mlflow/models/utils.py [hunk 7]
```diff
@@ -906,7 +906,7 @@ def _enforce_datatype(data: Any, dtype: DataType):
 def _enforce_array(data: Any, arr: Array, required=True):
     if not required and data is None:
         return None
-    if not isinstance(data, list):
+    if not isinstance(data, (list, np.ndarray)):
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/models/utils.py:909]: because pyspark converts an array column to a numpy-array column

File: tests/pyfunc/test_spark.py (hunk 6 above)
Comment (harupy) [reply tests/pyfunc/test_spark.py:1356]: Adding more tests

Comment (harupy) [reply tests/pyfunc/test_spark.py:1363]: Done!

File: mlflow/pyfunc/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/pyfunc/__init__.py]: Added!

**Review by serena-ruan — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
@serena-ruan Can we remove the failed numpy tests?

```
______________________ test_enforce_property_with_errors _______________________

    def test_enforce_property_with_errors():
        with pytest.raises(
            MlflowException, match=r"Failed to enforce schema of data `123` with dtype `string`"
        ):
            _enforce_property(123, Property("a", DataType.string))
    
        with pytest.raises(MlflowException, match=r"Expected data to be list, got ndarray"):
>           _enforce_property(
                np.array(["some_sentence1", "some_sentence2"]), Property("a", Array(DataType.string))
            )
E           Failed: DID NOT RAISE <class 'mlflow.exceptions.MlflowException'>
```

**Comment by serena-ruan** [comment]
> @serena-ruan Can we remove the failed numpy tests?
> 
> ```
> ______________________ test_enforce_property_with_errors _______________________
> 
>     def test_enforce_property_with_errors():
>         with pytest.raises(
>             MlflowException, match=r"Failed to enforce schema of data `123` with dtype `string`"
>         ):
>             _enforce_property(123, Property("a", DataType.string))
>     
>         with pytest.raises(MlflowException, match=r"Expected data to be list, got ndarray"):
> >           _enforce_property(
>                 np.array(["some_sentence1", "some_sentence2"]), Property("a", Array(DataType.string))
>             )
> E           Failed: DID NOT RAISE <class 'mlflow.exceptions.MlflowException'>
> ```

Yes! Maybe we should add a test of validating it's correct instead?

**Comment by harupy** [comment]
@serena-ruan updated the ndarray tests 🤖

**Pull request merged**